### PR TITLE
Simple nucleous

### PR DESCRIPTION
**Pull request by Eugene-hu**
Pretty big update: New ways to use bittensor API, new miner template, weights and bias integration, implementation of GPU training, bittensor axon fix, and removal of the reload function. Will need to update some of the tests, as lots of stuff was taken out.

### CODE REVIEW

File: bittensor/_dendrite/dendrite_impl.py [hunk 1]
```diff
@@ -452,7 +452,7 @@ def forward_text(
             raise ValueError(error_msg)
             
         # Check list types.
-        if not isinstance(inputs[0], torch.LongTensor):
+        if not ( isinstance(inputs[0], torch.cuda.LongTensor) or isinstance(inputs[0], torch.LongTensor)) :
```
Comment (unconst) [inline bittensor/_dendrite/dendrite_impl.py:455]: Why the cuda catch here?

Comment (unconst) [reply bittensor/_dendrite/dendrite_impl.py:455]: Does this not catch cuda tensors?

File: miners/text/template_miner.py [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+
+        # ---- Join based on weights ----
+        joining_weights = F.softmax( topk_weights, dim = 0 )
+        output = torch.zeros( (inputs.shape[0], inputs.shape[1], bittensor.__network_dim__)).to( self.config.miner.device )
+        for index, response in enumerate( responses ): 
+            output += response.to( self.config.miner.device ) * joining_weights[ topk_uids[index] ]
+
+        # ---- Punish peers with non-successful return ops ----
+        with torch.no_grad():
+            self.chain_weights[topk_uids[(return_ops != 0)]] -=  self.config.miner.punishment
```
Comment (unconst) [inline miners/text/template_miner.py]: This arg should be part of the nucleus I think.

File: miners/text/template_miner.py [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+        parser.add_argument('--miner.batch_size_train', type=int, help='Training batch size.', default=2)
+        parser.add_argument('--miner.restart_on_failure',  action='store_true', help='''Restart miner on unknown error.''', default=False)
+        parser.add_argument('--miner.compute_remote_gradients', action='store_true', help='''Does the miner compute and return gradients from backward queries.''', default=False)
+        parser.add_argument('--miner.accumulate_remote_gradients', action='store_true', help='''Does the miner accumulate remote gradients from backward queries.''', default=False)
+        parser.add_argument('--miner.n_topk_chain_weights', type=int, help='Maximum number of weights to submit to chain', default=100 )
+        parser.add_argument('--miner.name', type=str, help='Trials for this miner go in miner.root / (wallet_cold - wallet_hot) / miner.name ', default='gpt2_exodus')
+        parser.add_argument('--miner.device', type=str, help='miner default training device cpu/cuda', default=("cuda" if torch.cuda.is_available() else "cpu"))
+        parser.add_argument('--miner.punishment', type=int, help='The punishment on responses that do no ', default=100 )
```
Comment (unconst) [inline miners/text/template_miner.py]: Needs to be complete sentence.

File: miners/text/template_miner.py [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+    def checkpoint( self ):
+        r""" Optionally Saves, updates and then reloads the miner training state.
+        """
+        last_saved = self.get_saved_state()
+        if last_saved == None or last_saved['epoch_loss'] >= self.epoch_loss:
+            self.save()
+        bittensor.neuron.metagraph.load()
+        bittensor.neuron.metagraph.sync()
+        bittensor.neuron.metagraph.save()
+        #self.reload()
```
Comment (unconst) [inline miners/text/template_miner.py]: We should add a NAN check here, something like "have we nan'd out? yes = reload()"

File: miners/text/template_miner.py [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+        # --- loads and syncs metagraph
+        bittensor.neuron.metagraph.load()
+        bittensor.neuron.metagraph.sync()
+        bittensor.neuron.metagraph.save()
+
+        # ---- Load training state.
+        self.epoch = state_dict['epoch']
+        self.epoch_loss = state_dict['epoch_loss']
+        self.global_step = state_dict['global_step']
+        if 'network' in state_dict.keys() and bittensor.neuron.subtensor.network == state_dict['network']: # checks if you are loading into the same network
```
Comment (unconst) [inline miners/text/template_miner.py:542]: I think we should probably just error out here. This should basically never happen.

**Review by unconst — APPROVED** [review]

File: bittensor/_dendrite/dendrite_impl.py (hunk 1 above)
Comment (Eugene-hu) [reply bittensor/_dendrite/dendrite_impl.py:455]: Yea, so this was one of the main issues, I ran into. In torch, cuda tensors and cpu have different instances and need to be separated.

File: miners/text/template_miner.py (hunk 2 above)
Comment (Eugene-hu) [reply miners/text/template_miner.py]: Fixed!

File: miners/text/template_miner.py (hunk 4 above)
Comment (Eugene-hu) [reply miners/text/template_miner.py]: Fixed

File: miners/text/template_miner.py (hunk 5 above)
Comment (Eugene-hu) [reply miners/text/template_miner.py:542]: Fixed

**Review by shibshib — APPROVED** [review]

**Pull request merged**